Realm/RLMApp.mm

```diff
@@ -93,119 +98,159 @@ - (instancetype)initWithBaseURL:(nullable NSString *) baseURL
+}
+
 @end
 
 NSError *RLMAppErrorToNSError(realm::app::AppError const& appError) {
     return [[NSError alloc] initWithDomain:@(appError.error_code.category().name())
                                       code:appError.error_code.value()
                                   userInfo:@{
-                                      @(appError.error_code.category().name()) : @(appError.error_code.message().data())
+                                      @(appError.error_code.category().name()) : @(appError.error_code.message().data()),
+                                      NSLocalizedDescriptionKey : @(appError.message.c_str())
                                   }];
 }
 
 @interface RLMApp() {
-    realm::app::App _app;
+    std::shared_ptr<realm::app::App> _app;
 }
+
 @end
 
 @implementation RLMApp : NSObject
 
-- (instancetype)initWithAppId:(NSString *)appId configuration:(RLMAppConfiguration *)configuration {
+static NSMutableDictionary *_apps = [NSMutableDictionary new];
+
++ (NSDictionary<NSString *,RLMApp *> *)apps {
+    return _apps;
+}
+
+- (instancetype)initWithAppId:(NSString *)appId
+                configuration:(RLMAppConfiguration *)configuration
+                rootDirectory:(NSURL *)rootDirectory {
     if (self = [super init]) {
-        app::App::Config boundConfiguration = {
-            .app_id = appId.UTF8String
-        };
-        boundConfiguration.transport_generator = [configuration]{
-            return std::make_unique<CocoaNetworkTransport>([RLMNetworkTransport new]);
-        };
-        if (configuration) {
-            if (configuration.baseURL) {
-                boundConfiguration.base_url = util::Optional<std::string>(configuration.baseURL.UTF8String);
-            }
-            if (configuration.transport) {
-                boundConfiguration.transport_generator = [configuration]{
-                    return std::make_unique<CocoaNetworkTransport>(configuration.transport);
-                };
-            }
-            if (configuration.localAppName) {
-                boundConfiguration.local_app_name = std::string(configuration.localAppName.UTF8String);
-            }
-            if (configuration.localAppVersion) {
-                boundConfiguration.local_app_version = std::string(configuration.localAppVersion.UTF8String);
-            }
-            boundConfiguration.default_request_timeout_ms = (uint64_t)configuration.defaultRequestTimeoutMS;
-        }
-        _app = realm::app::App(boundConfiguration);
+        _configuration = configuration;
+        [_configuration setAppId: appId];
+
+        _syncManager = [[RLMSyncManager alloc] initWithAppConfiguration:configuration rootDirectory:rootDirectory];
+        _app = [_syncManager app];
+
         return self;
     }
     return nil;
 }
 
++ (instancetype)app:(NSString *)appId
+      configuration:(RLMAppConfiguration *)configuration
+      rootDirectory:(NSURL *)rootDirectory {
+    if (RLMApp *app = _apps[appId]) {
```

Comment:
This method (and everything else accessing `_apps`) needs locking. Rather than exposing the dictionary directly there should be a lookup function which returns the app for an id (with the lookup done with a lock held).